src/ripple/shamap/impl/ShardFamily.cpp

```diff
@@ -0,0 +1,189 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/ledger/LedgerMaster.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/app/main/Tuning.h>
+#include <ripple/nodestore/DatabaseShard.h>
+#include <ripple/shamap/ShardFamily.h>
+
+namespace ripple {
+
+ShardFamily::ShardFamily(Application& app, CollectorManager& cm)
+    : app_(app)
+    , db_(*app.getShardStore())
+    , cm_(cm)
+    , j_(app.journal("ShardFamily"))
+    , tnTargetSize_(app.config().getValueFor(SizedItem::treeCacheSize))
+    , tnTargetAge_(app.config().getValueFor(SizedItem::treeCacheAge))
+{
+    assert(app.getShardStore());
```

Comment:
If you hit this `assert` I think you already had undefined behavior, since you assigned the contents of a `nullptr` to a reference.  So all bets were already off.  The `assert` may or may not fire.

Since you're in a .cpp file, you can work around this with the following trick...
```
static NodeStore::Database&
getShardStore (Application& app)
{
    auto const dbPtr = app.getShardStore();
    assert (dbPtr);
    return *dbPtr;
}

ShardFamily::ShardFamily(Application& app, CollectorManager& cm)
    : app_(app)
    , db_(getShardStore(app))
    , cm_(cm)
    , j_(app.journal("ShardFamily"))
    , tnTargetSize_(app.config().getValueFor(SizedItem::treeCacheSize))
    , tnTargetAge_(app.config().getValueFor(SizedItem::treeCacheAge))
{
}
```